From ae2f72d695c85b889c6cd8a9c7850c133761e205 Mon Sep 17 00:00:00 2001
From: Zhan Lou <zhan.lou@cixtech.com>
Date: Thu, 25 Sep 2025 19:43:42 +0800
Subject: [PATCH] DPTSW-19924: v4l2: fix seek hang issue

The v4l2bufferpool is locked before gst_v4l2_buffer_pool_streamoff(),
and later on it is locked again in gst_v4l2_buffer_pool_qbuf() which is
called in gst_v4l2_buffer_pool_streamoff(). This introduces deadlock.

Solution: unlock v4l2bufferpool before calling into
gst_v4l2_buffer_pool_qbuf(), and lock again after qbuf is done.

Signed-off-by: Zhan Lou <zhan.lou@cixtech.com>
Change-Id: If4bcce8a93e920393b087cfdecb517089651390e

diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index 694f8a0e58..1a9dff78cb 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -836,11 +836,16 @@ gst_v4l2_buffer_pool_streamoff (GstV4l2BufferPool * pool)
       pool->buffers[i] = NULL;
 
       if (!(buffer_state[i] & BUFFER_STATE_OUTSTANDING)) {
+        /* Unlock since gst_v4l2_buffer_pool_qbuf() will lock the pool */
+        GST_OBJECT_UNLOCK (pool);
+
         if (V4L2_TYPE_IS_OUTPUT (pool->obj->type))
           gst_v4l2_buffer_pool_complete_release_buffer (bpool, buffer, FALSE);
 
         else                    /* Don't re-enqueue capture buffer on stop */
           pclass->release_buffer (bpool, buffer);
+
+        GST_OBJECT_LOCK (pool);
       }
 
       g_atomic_int_add (&pool->num_queued, -1);
@@ -1102,8 +1107,11 @@ gst_v4l2_buffer_pool_stop (GstBufferPool * bpool)
     pool->other_pool = NULL;
   }
 
-  if (!pool->orphaned)
+  if (!pool->orphaned) {
+    GST_OBJECT_LOCK (pool);
     gst_v4l2_buffer_pool_streamoff (pool);
+    GST_OBJECT_UNLOCK (pool);
+  }
 
   ret = GST_BUFFER_POOL_CLASS (parent_class)->stop (bpool);
 
